### PR TITLE
blog: Extensions

### PR DESCRIPTION
**Pull request by victorwuky**
Will merge (and thus publish blog post) after:

- Reviews here.
- Victor has tested the bug fix with the latest Tilt release: https://app.clubhouse.io/windmill/story/4708/extensions-are-executed-with-a-working-directory-in-the-tilt-modules-dir
  - Done.

screencapture-localhost-4002-2020-04-01-more-customizable-tiltfiles-with-extensions-html-2020-04-01-15_00_58

### CODE REVIEW

**Comment by victorwuky** [comment]
@nicks : Can you please review? Especially the thing about free t-shirt at the end.

File: blog/_posts/2020-03-31-extensions.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  - tilt
+  - tiltfile
+---
+
+Since [releasing Tilt](/2018/08/28/how-tilt-updates-kubernetes-in-seconds-not-minutes.html), we’ve seen the Tilt community actively share code snippets of Tiltfile functionality in [Slack #tilt](https://kubernetes.slack.com/messages/CESBL84MV/), including ideas such as forcing resources into different namespaces, injecting sidecars, and even running tests. So we've created the open source Extensions platform to encourage this community collaboration in more structured fashion.
+
+Tilt Extensions is available starting in v0.12.11. [Upgrade](https://docs.tilt.dev/upgrade.html) to the latest version.
+
+## What's an extension, exactly? And how do you use it?
+An extension is simply a packaged function you call from within your Tiltfile. Simply import the extension (specifying the extension name and function name), and call it, all right inside your Tiltfile.
```
Comment (nicks) [inline blog/_posts/2020-03-31-extensions.md]: We use the word "simple" a lot in this blog post. Let's try to avoid it. For the audience we're writing for (infra engineers), they're going to be sensitive to overuse of words like "easy", "simple", "just", "straightforward", etc.

The old cliche of "show, don't tell" applies here.

One way to address this is to explain what we're trying to optimize for. e.g.,

```
You may have already defined a function in your Tiltfile!

def hi():
  print("hello world")

With Tilt Extensions, we wanted to create a clear path from:

- Take an existing a function in your Tiltfile
- Publish that function to a shared repository
- Replace the private function with the published version
````

File: blog/_posts/2020-03-31-extensions.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+image_caption: "Photo by <a href='https://unsplash.com/@lazycreekimages?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText'>Michael Dziedzic</a> on <a href='https://unsplash.com/?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText'>Unsplash</a>"
+tags:
+  - tilt  
+keywords:
+  - extensions
+  - tilt
+  - tiltfile
+---
+
+Since [releasing Tilt](/2018/08/28/how-tilt-updates-kubernetes-in-seconds-not-minutes.html), we’ve seen the Tilt community actively share code snippets of Tiltfile functionality in [Slack #tilt](https://kubernetes.slack.com/messages/CESBL84MV/), including ideas such as forcing resources into different namespaces, injecting sidecars, and even running tests. So we've created the open source Extensions platform to encourage this community collaboration in more structured fashion.
```
Comment (nicks) [inline blog/_posts/2020-03-31-extensions.md]: I feel like this dumps a ton context on the reader. In the first sentence, it jumps around in time, and introduces several features, without a lot of guidance and what to focus on. I have a hard time parsing it, and I'm pretty familiar with Tilt :)

```
Tilt lets you declare your dev environment in Starlark, a subset of Python.

But Starlark is missing a big feature! There's no package manager. Starlark doesn't have built-in ways to share your configuration.

The community noticed! We saw people:
- copy and paste "best practices" for different languages
- struggle to share Tiltfiles across multiple repos
- send each other Tiltfile snippets

Tilt extensions are a first baby step towards a multi-service dev package manager.
```

and maybe at the bottom of the post plug the types of extensions we're hoping to see?

[the remark this answers is not in the document]
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: I linked to the medium blog instead. Let me know if you prefer just not linking it doing something different.

File: blog/_posts/2020-03-31-extensions.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+```
+
+Then call `hi()` later in your Tiltfile to print "Hello World!".
+
+Read the [docs](https://docs.tilt.dev/extensions.html) for additional details and look at the [tilt-extensions repo](https://github.com/windmilleng/tilt-extensions) for a list of available extensions. But it's really that simple!
+
+## I want to contribute an extension
+If you're interested in contributing an extension, it's really easy. You just need to know some basic Python syntax. And if you've had practice editing a Tiltfile, that's good enough. You definitely do not need to know Go (which Tilt is written in) in order to contribute. Follow the simple instructions in [Contribute an Extension](https://docs.tilt.dev/contribute_extension.html) to submit a pull request.
+
+If you submit an extension within the month of April 2020, and it's accepted and published, we'll send you a free Tilt t-shirt. [They look really awesome](https://twitter.com/tilt_dev/status/1212769783034384384)!
```
[the remark this answers is not in the document]
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: Fixed.

File: blog/_posts/2020-03-31-extensions.md (hunk 1 above)
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: > We use the word "simple" a lot in this blog post. Let's try to avoid it. For the audience we're writing for (infra engineers), they're going to be sensitive to overuse of words like "easy", "simple", "just", "straightforward", etc.

I've removed all "simple"s in the blog post. 

> The old cliche of "show, don't tell" applies here.

I've simplified the explanation and relied on showing the "load" command.

> One way to address this is to explain what we're trying to optimize for. e.g.,

I don't think this directly applies in this section, since it is about using extensions, not taking an existing function and then publishing it. But I do think it's good content. I've moved a version of this content to the subsequent section about contributing an extension.

File: blog/_posts/2020-03-31-extensions.md (hunk 2 above)
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: I added this to the top section and pretty much removed most of the original paragraph.

> and maybe at the bottom of the post plug the types of extensions we're hoping to see?

Which types of extensions are we hoping to see? I'm not sure "forcing resources into different namespaces, injecting sidecars, and even running tests" are necessarily things we want?

File: blog/_posts/2020-03-31-extensions.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  - tilt  
+keywords:
+  - extensions
+  - tilt
+  - tiltfile
+---
+
+Since [releasing Tilt](/2018/08/28/how-tilt-updates-kubernetes-in-seconds-not-minutes.html), we’ve seen the Tilt community actively share code snippets of Tiltfile functionality in [Slack #tilt](https://kubernetes.slack.com/messages/CESBL84MV/), including ideas such as forcing resources into different namespaces, injecting sidecars, and even running tests. So we've created the open source Extensions platform to encourage this community collaboration in more structured fashion.
+
+Tilt Extensions is available starting in v0.12.11. [Upgrade](https://docs.tilt.dev/upgrade.html) to the latest version.
```
[the remark this answers is not in the document]
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: This section has been replaced. In particular, there's no longer proper nouns for extensions anymore.

**Comment by victorwuky** [comment]
@nicks : I've incorporated your helpful feedback. Do you mind checking again? Thanks.

File: blog/_posts/2020-03-31-extensions.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+- Send each other Tiltfile snippets
+
+So we've created Tilt extensions as a first baby step toward a multi-service dev package manager to address this gap.
+
+Tilt extensions are available starting in [v0.12.11](https://github.com/windmilleng/tilt/releases/tag/v0.12.11). [Upgrade](https://docs.tilt.dev/upgrade.html) to the latest version of Tilt.
+
+## What's an extension, exactly? And how do you use it?
+An extension is a packaged function that you import and call, all right inside your Tiltfile. 
+
+For example, to use the [`hello_world` extension](https://github.com/windmilleng/tilt-extensions/tree/master/hello_world), first [`load()`](https://docs.tilt.dev/api.html#api.load) it by including this command in your Tiltfile:
```
Comment (landism) [inline blog/_posts/2020-03-31-extensions.md:32]:
```suggestion
For example, to use the `hi` function from the [`hello_world` extension](https://github.com/windmilleng/tilt-extensions/tree/master/hello_world), first [`load()`](https://docs.tilt.dev/api.html#api.load) it by including this command in your Tiltfile:
```

Comment (landism) [reply blog/_posts/2020-03-31-extensions.md:32]: just to make the second parameter in the example clearer

File: blog/_posts/2020-03-31-extensions.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+If you're interested in contributing an extension, you just need to know some basic Python syntax. And if you've had practice editing a Tiltfile, that's good enough. Knowledge of Go (which Tilt is written in) is not required in order to contribute. Follow the instructions in [Contribute an Extension](https://docs.tilt.dev/contribute_extension.html) to submit a pull request.
+
+If you've already defined a function in your Tiltfile, we've designed extensions to create a clear path for you to:
+
+- Take an existing function in your Tiltfile
+- Publish that function to the [tilt-extensions repo](https://github.com/windmilleng/tilt-extensions)
+- Replace that function in your Tiltfile with the published version
+
+## Get a free t-shirt
+If you submit an extension within the month of April 2020, and it's accepted and published, we'll send you a free Tilt t-shirt (as long as you're in the United States, to simplify shipping). [They look really awesome!](https://twitter.com/tilt_dev/status/1212769783034384384)
```
Comment (landism) [inline blog/_posts/2020-03-31-extensions.md:52]: Is it worth limiting the # of shirts we're committing to to like the first N, or the best N at the end of the month, or is T-shirt budget not a concern?

Comment (ellenkorbes) [reply blog/_posts/2020-03-31-extensions.md:52]: Shouldn't be a concern.

File: blog/_posts/2020-03-31-extensions.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+image_caption: "Photo by <a href='https://unsplash.com/@lazycreekimages?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText'>Michael Dziedzic</a> on <a href='https://unsplash.com/?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText'>Unsplash</a>"
+tags:
+  - tilt
+keywords:
+  - extensions
+  - tilt
+  - tiltfile
+---
+
+Tilt lets you declare your dev environment in Starlark, a subset of Python. But Starlark is missing a big feature! There's no package manager. Starlark doesn't have built-in ways to share your configuration.
```
Comment (ellenkorbes) [inline blog/_posts/2020-03-31-extensions.md:18]: Let's open with an active voice statement.

```suggestion
In Tilt, you declare your dev environment in Starlark, a subset of Python. But Starlark is missing a big feature: there's no package manager! Starlark doesn't have built-in ways to share your configuration.
```

One more thing, I find that the word `configuration` sounds really weird to me. It doesn't tell me what it is that I'd like to share. Am I sharing a file? A function? What is the tangible item being shared?

File: blog/_posts/2020-03-31-extensions.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+The Tilt community noticed! We've seen people:
+- Copy and paste "best practices" for different languages
+- Struggle to share Tiltfiles across multiple repos
+- Send each other Tiltfile snippets
+
+So we've created Tilt extensions as a first baby step toward a multi-service dev package manager to address this gap.
```
Comment (ellenkorbes) [inline blog/_posts/2020-03-31-extensions.md:25]:
```suggestion
To address this gap, as a first step toward a multi-service dev package manager, we've created Tilt extensions.
```

File: blog/_posts/2020-03-31-extensions.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+Tilt extensions are available starting in [v0.12.11](https://github.com/windmilleng/tilt/releases/tag/v0.12.11). [Upgrade](https://docs.tilt.dev/upgrade.html) to the latest version of Tilt.
```
Comment (ellenkorbes) [inline blog/_posts/2020-03-31-extensions.md:27]:
```suggestion
Tilt extensions are available starting in [v0.12.11](https://github.com/windmilleng/tilt/releases/tag/v0.12.11). ([Click here](https://docs.tilt.dev/upgrade.html) to upgrade!)
```

File: blog/_posts/2020-03-31-extensions.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+## What's an extension, exactly? And how do you use it?
+An extension is a packaged function that you import and call, all right inside your Tiltfile.
```
Comment (ellenkorbes) [inline blog/_posts/2020-03-31-extensions.md:30]: You can also import another file locally using `load()`, "right inside your Tiltfile." That's not the point. The point is to do so remotely, and more importantly, to be able to use functionality you didn't write yourself (since ideally a lot of people will be sharing these.)

Without making these points, the 14-year-old edgelord nerd in me reads this whole section and replies with, "and why should I care?"

File: blog/_posts/2020-03-31-extensions.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+For example, to use the [`hello_world` extension](https://github.com/windmilleng/tilt-extensions/tree/master/hello_world), first [`load()`](https://docs.tilt.dev/api.html#api.load) it by including this command in your Tiltfile:
+
+```
+load('ext://hello_world', 'hi')
+```
+
+Then call [`hi()`](https://github.com/windmilleng/tilt-extensions/tree/master/hello_world/Tiltfile) later in your Tiltfile to print "Hello World!".
+
+Read the [docs](https://docs.tilt.dev/extensions.html) for additional details and look at the [tilt-extensions repo](https://github.com/windmilleng/tilt-extensions) for a list of available extensions.
```
Comment (ellenkorbes) [inline blog/_posts/2020-03-31-extensions.md:40]: I'd love to see a few words before the "please contribute" CTA about where we see this going.

What types of extensions are we expecting to see? Why should I be excited to contribute? What do I get out of investing my time in this? Is this a community, an ecosystem?

File: blog/_posts/2020-03-31-extensions.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+## I want to contribute an extension
+If you're interested in contributing an extension, you just need to know some basic Python syntax. And if you've had practice editing a Tiltfile, that's good enough. Knowledge of Go (which Tilt is written in) is not required in order to contribute. Follow the instructions in [Contribute an Extension](https://docs.tilt.dev/contribute_extension.html) to submit a pull request.
```
Comment (ellenkorbes) [inline blog/_posts/2020-03-31-extensions.md:43]: I'm not a fan of the tone in the first two sentences—"you just need to know..." To me in conveys a false sense of simplicity, when in reality there is finesse involved. Python is simple but writing good Python is not. I suggest rephrasing.

File: blog/_posts/2020-03-31-extensions.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+If you've already defined a function in your Tiltfile, we've designed extensions to create a clear path for you to:
+
+- Take an existing function in your Tiltfile
+- Publish that function to the [tilt-extensions repo](https://github.com/windmilleng/tilt-extensions)
+- Replace that function in your Tiltfile with the published version
+
+## Get a free t-shirt
```
Comment (ellenkorbes) [inline blog/_posts/2020-03-31-extensions.md:51]: The word `free` conveys a sense of it being a financial transaction, "do [thing] and look what you'll get in return!" But no one's gonna write an extension just because they want the t-shirt.

What's really going on is us wanting to express gratitude, to show contributors their efforts are appreciated. I suggest rewording with a more "from the heart" tone—it'll also more accurately reflects the general attitude within the company.

File: blog/_posts/2020-03-31-extensions.md (hunk 5 above)
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md:32]: Thanks @landism . Committing your suggestion.

File: blog/_posts/2020-03-31-extensions.md (hunk 6 above)
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md:52]: I don't think this is a concern either. If we get too many PRs, that's a good thing. Let's hope for a t-shirt budget problem!

File: blog/_posts/2020-03-31-extensions.md (hunk 7 above)
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: I updated to be "Starlark doesn't have built-in ways to share code."

File: blog/_posts/2020-03-31-extensions.md (hunk 12 above)
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: I rephrased this so it's more matter-of-fact, and doesn't make subjective statements.

File: blog/_posts/2020-03-31-extensions.md (hunk 13 above)
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: I re-worded and removed all instances of "free" in the post.

File: blog/_posts/2020-03-31-extensions.md (hunk 10 above)
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: Re-worded to includes these two points. Great observation.

File: blog/_posts/2020-03-31-extensions.md (hunk 11 above)
Comment (victorwuky) [reply blog/_posts/2020-03-31-extensions.md]: Updated with vision of future and also added expected extensions in contributing section.

**Comment by victorwuky** [comment]
@ellenkorbes : Merging. Feel free to let me know if you have other changes you want or just create a PR directly and tag me. Happy to adjust going forward.
